Working cl_khr_icd 2.0.0.

Kerilk opened the pull request:
This implements the loader managed dispatch strategy presented here: https://github.com/KhronosGroup/OpenCL-Docs/issues/1003

pjaaskel commented on lib/CL/devices/devices.c:
```diff
@@ -319,24 +320,25 @@ pocl_get_devices (cl_device_type device_type, cl_device_id *devices,
               {
-                devices[dev_added] = &pocl_devices[i];
-                ++dev_added;
+              devices[dev_added] = platform->devices[i];
```
Is there a tab vs. WS difference here? Pls run format-{diff,branch}.sh

Kerilk replied:
format-diff.sh was used. I elected not to apply changes to some macro definitions, indentation was being lost.

pjaaskel commented on lib/CL/devices/devices.h:
```diff
@@ -74,6 +81,25 @@ unsigned int pocl_get_devices(cl_device_type device_type, struct _cl_device_id *
+void pocl_set_devices_dispatch_data (cl_platform_id platform, void *disp_data);
+
+/**
+ * \brief get a instance copy of pocl_devices
```
Please elaborate in the comment what an "instance copy" means.

Kerilk replied:
Done.

pjaaskel commented on lib/CL/clCreateContext.c:
```diff
@@ -150,20 +171,23 @@ POname(clCreateContext)(const cl_context_properties * properties,
-    errcode = CL_INVALID_DEVICE;
-  POCL_GOTO_ERROR_ON ((errcode != CL_SUCCESS), errcode,
-                      "Could not initialize devices\n");
+  if (!platform->instance)
```
I prefer to be explicit with the predicates: "platform->instance != NULL" instead of relying on the implicit boolean conversion.

Kerilk replied:
Instance are not part of the PR, to reflect what was merged on the Khronos github.
Will reopen a PR with the instance when this is merged.

pjaaskel commented on lib/CL/clIcdCreateInstancePlatformKHR.c:
```diff
@@ -0,0 +1,67 @@
+/* OpenCL runtime library: clSetPlatformDispatchDataKHR()
+
+   Copyright (c) 2023 Brice Videau / Argonne National Laboratory
```
Should this be 2023-2024?

Kerilk replied:
addressed.

pjaaskel commented on lib/CL/clGetPlatformIDs.c:
```diff
@@ -37,7 +37,7 @@
 #ifdef BUILD_ICD
 struct _cl_icd_dispatch pocl_dispatch = {
-  &POname(clGetPlatformIDs),
+  (void *)CL_ICD2_TAG_KHR,
```
Pls add a comment explaining what the TAG does (so the code reader doesn't have to check the ICD2 spec).

Kerilk replied:
Addressed.

pjaaskel commented on lib/CL/clIcdCreateInstancePlatformKHR.c:
```diff
@@ -0,0 +1,67 @@
+ERROR:
+  POCL_MEM_FREE (new_platform);
+  if (errcode_ret)
+    {
```
brackets not needed

Kerilk replied:
Instance are not part of the PR, to reflect what was merged on the Khronos github.
Will reopen a PR with the instance when this is merged.

pjaaskel commented on lib/CL/devices/devices.c:
```diff
@@ -668,11 +677,71 @@ pocl_init_devices ()
 #endif
-
+  for (unsigned i = 0; i < pocl_num_devices; i++)
+    {
```
brackets not needed

Kerilk replied:
Addressed.

pjaaskel commented on lib/CL/devices/devices.c:
```diff
@@ -668,11 +677,71 @@ pocl_init_devices ()
+pocl_set_devices_dispatch_data (cl_platform_id platform, void *disp_data)
+{
+  for (unsigned int i = 0; i < platform->num_devices; i++)
+    {
```
extra brackets

Kerilk replied:
Addressed.

pjaaskel commented on lib/CL/pocl_cl.h:
```diff
@@ -286,16 +287,55 @@ extern uint64_t last_object_id;
 /* The ICD compatibility part. This must be first in the objects where
  * it is used (as the ICD loader assumes that)*/
+#ifndef CL_ICD2_TAG_KHR
+#define CL_ICD2_TAG_KHR ((size_t)0x4F50454E434C3331ULL)
```
Perhaps add the comment here. A magic number to detect we are using ICD2?

Kerilk replied:
Addressed.

Kerilk commented:
I have rebased the PR on `main`, and removed the instance part, to reflect what was merged in the official Khronos repository. I will reopen a new PR with instance support once this is merged.

Code is tested and functional on my laptop, and the resulting pocl icd loads with both up to date or outdated loaders using the icd 2 or icd 1 interface respectively.

I have addressed the comments and left open those that may raise discussions.

Kerilk commented:
Pushed an update to address ICD on windows, function declaration was missing `CL_API_ENTRY`.

Kerilk commented:
Fixed setting device dispatch data was not traversing the linked list correctly.

Kerilk commented:
Fix global variable naming in network discovery.

Kerilk commented:
Fix `pocl_##__DRV__##_init_device_discovery` prototype in `prototypes.inc`. Thanks for the CI.

Kerilk commented:
> @Kerilk FYI i think all of the remaining errors in the CIs (asan/tsan, tier1 & remote) will be fixed by pending PR #1864, you can ignore them.

Thanks, once yours is merged, I'll rebase, so I can sleep better :)

Kerilk commented:
Rebased on main, which should fix failing CI tests.
